### PR TITLE
Keyimport2

### PR DESCRIPTION
**Pull request by hodarh**

### CODE REVIEW

**Comment by hodarh** [comment]
#build

**Comment by hodarh** [comment]
#build

**Comment by hodarh** [comment]
#build

File: enforcer/src/enforcer/enforcer.c [hunk 1]
```diff
@@ -2582,13 +2515,12 @@ updatePolicy(engine_type *engine, db_connection_t *dbconn, policy_t *policy,
...
-		tag = keytag(hsm_key_locator(hsmkey), hsm_key_algorithm(hsmkey),
+		success = hsm_keytag(hsm_key_locator(hsmkey), hsm_key_algorithm(hsmkey),
 			((hsm_key_role(hsmkey) == HSM_KEY_ROLE_KSK
 				|| hsm_key_role(hsmkey) == HSM_KEY_ROLE_CSK)
 				? 1 : 0),
-			&success);
-		if (!success
-			|| key_data_set_keytag(mutkey, tag))
+			&tag);
+		if (success || key_data_set_keytag(mutkey, tag))
```
Comment (yschaeff) [inline enforcer/src/enforcer/enforcer.c]: This lines should conflict on merge. If they don't please review result.

File: enforcer/src/keystate/keystate_import_cmd.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2016 Surfnet 
+ * Copyright (c) 2016 .SE (The Internet Infrastructure Foundation).
+ * Copyright (c) 2016 OpenDNSSEC AB (svb)
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.h]: * Copyright (c) 2016 NLNet Labs. All rights reserved.

instead?

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+	client_printf_err(sockfd, "Unable to find the key with this locator: %s\n", ckaid);
+	hsm_destroy_context(hsm_ctx);
+	return -1;
+    }
+
+    hsm_key = hsm_key_new_get_by_locator(dbconn, ckaid);
+    if (hsm_key) {
+        ods_log_error("[%s] Error: Already used this key with this locator: %s", module_str, ckaid);
+        client_printf_err(sockfd, "Already used this key with this locator: %s\n", ckaid);
+        return -1;
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: hsm_key_free(hsm_key); first?

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    }
+
+    if (!hsm_find_key_by_id(hsm_ctx, ckaid)) {
+	ods_log_error("[%s] Error: Unable to find the key with this locator: %s", module_str, ckaid);
+	client_printf_err(sockfd, "Unable to find the key with this locator: %s\n", ckaid);
+	hsm_destroy_context(hsm_ctx);
+	return -1;
+    }
+
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: else {
  free(the thing from hsm_find_key_by_id())
}

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
+    if (!hsm_find_key_by_id(hsm_ctx, ckaid)) {
+        ods_log_error("[%s] Error: Unable to find the key with this locator: %s", module_str, ckaid);
+        client_printf_err(sockfd, "Unable to find the key with this locator: %s\n", ckaid);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: else {
free(the thing from hsm_find_key_by_id())
}

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+		|| key_data_set_role(key_data, keytype)
+		|| key_data_set_ds_at_parent(key_data, keytype == 1 ? ds_at_parent[keystate] : 0)
+		|| key_data_set_keytag(key_data, tag)
+		|| key_data_set_minimize(key_data,setmin)
+                || key_data_create(key_data))
+    {
+        ods_log_error("[%s] key data creation failed, database or memory error", module_str);
+        key_data_free(key_data);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: hsm_key_free(hsmkey); ?

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+        }
+        else {
+            ods_log_error("[%s] Error: Unable to find repository %s in HSM", module_str, rep);
+            client_printf_err(sockfd, "Unable to find repository %s in HSM\n", rep);
+        }
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
+    if (!hsm_find_key_by_id(hsm_ctx, ckaid)) {
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: free hsm_find_key_by_id() result

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+                || key_state_set_ttl (key_state, policy_parent_ds_ttl(policy))
+		|| key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][0] : zsk_mapping[keystate][0])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for DS failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: key_state_free(key_state);

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+                || key_state_set_ttl (key_state, policy_keys_ttl(policy))
+		|| key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][1] : zsk_mapping[keystate][1])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for DNSKEY failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: key_state_free(key_state);

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+                || key_state_set_ttl (key_state, policy_keys_ttl(policy))
+                || key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][2] : zsk_mapping[keystate][2])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for RRSIGDNSKEY failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: key_state_free(key_state);

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+    ttl = max(min(policy_zone_soa_ttl(policy), policy_zone_soa_minimum(policy)),
+            (policy_denial_type(policy) == POLICY_DENIAL_TYPE_NSEC3
+                ? ( policy_denial_ttl(policy) > policy_signatures_max_zone_ttl(policy)
+                    ? policy_denial_ttl(policy)
+                    : policy_signatures_max_zone_ttl(policy))
+                : policy_signatures_max_zone_ttl(policy)));
+
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: This is horrible. (I see you copied it from enforcer.c, not blaming you.) We should consider rewriting this to something readable...

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    client_printf(sockfd,
+		"key import             Import DNSKEY(s) for a given zone to the database.\n"
+		"      --cka_id <CKA_ID>		(aka -k)  cka_id\n"
+		"      --repository <repository>	(aka -r) repository\n"
+		"      --zone <zone>			(aka -z)  zone \n"
+		"      --bits <size>			(aka -b) bits\n"
+		"      --algorithm <algorithm>		(aka -g) algorithm \n"
+		"      --keystate <state>		(aka -e) keystate\n"
+		"      --keytype <type>			(aka -t) keytype\n"
+		"      --time <time>			(aka -w) time\n"
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: I believe just "time" doesn't explain enough in this context.

File: enforcer/src/keystate/keystate_import_cmd.c [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+#include "libhsmdns.h"
+#include "db/key_data.h"
+#include "db/db_error.h"
+
+#include "keystate/keystate_import_cmd.h"
+#include "keystate/keystate_list_cmd.h"
+
+static const char *module_str = "keystate_import_cmd";
+
+const int ksk_mapping[5][4] = {{0,0,0,4},{0,1,1,4},{1,2,2,4},{1,2,2,4},{3,2,2,4}};
```
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: could you add a comment on what the columns and rows mean?

File: enforcer/src/keystate/keystate_import_cmd.c (hunk 12 above)
Comment (hodarh) [inline enforcer/src/keystate/keystate_import_cmd.c]: thanks for review, I used 1.4 format. inception time? is it good or have any suggestion?

**Comment by hodarh** [comment]
#build

File: enforcer/src/keystate/keystate_import_cmd.c (hunk 12 above)
Comment (yschaeff) [inline enforcer/src/keystate/keystate_import_cmd.c]: "Inception time" is already much better yes.

**Comment by hodarh** [comment]
#build

**Pull request merged**